Support CTE/JOIN usage for PIVOT operator

RIC06X opened the pull request:
### Description

This commit fixed an issue that CTE and JOIN cannot be used with pivot operator.

We added 1 field (`Node *context`) in FuncExpr Struct to store pivot additional information. FuncExpr will carry the pivot information to ExecMakeTableFunctionResult. In this function, we pass the pivot information into two newly added FunctionCallInfo that we can retrieve at bbf_pivot function.  

### Issues Resolved

Task: BABEL-4713, BABEL-4714, BABEL-4795
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

RIC06X committed:
[BABEL-4673] Support View usage for PIVOT operator

This commit fixed an issue that view cannot be created or used by stmt
with pivot operator.

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

xhfanhe commented on src/include/nodes/parsenodes.h:
```diff
@@ -235,6 +235,9 @@ typedef struct Query
 	/* length in bytes; 0 means "rest of string" */
 	int			stmt_len pg_node_attr(query_jumble_ignore);
+
+	bool		isPivot pg_node_attr(query_jumble_ignore);
+	List		*pivotInfoList pg_node_attr(query_jumble_ignore);
```
I feel this causes too many engine changes... but if it's a must, so be it. Please add comments to explain the change.

xhfanhe commented on src/backend/parser/analyze.c:
```diff
@@ -407,11 +407,23 @@ transformStmt(ParseState *pstate, Node *parseTree)
 				SelectStmt *n = (SelectStmt *) parseTree;
 
 				if (n->valuesLists)
+				{
 					result = transformValuesClause(pstate, n);
+					result->isPivot = false;	
+					result->pivotInfoList = NIL;
+				}
+					
 				else if (n->op == SETOP_NONE)
+				{
 					result = transformSelectStmt(pstate, n);
+				}
 				else
+				{
 					result = transformSetOperationStmt(pstate, n);
+					result->isPivot = false;	
+					result->pivotInfoList = NIL;
+				}
```
Nit: this can be simplified as 
```
result->isPivot = false;
result->pivotInfoList = NIL;

if (...)
```

RIC06X committed:
Add CTE/VIEW/JOIN support for PIVOT

Modified internal PIVOT information analyzer -> executor passing design.
Supported CTE/VIEW/JOIN as well as function/procedure for PIVOT
operator.

Task: BABEL-4713, BABEL-4714, BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4673] Disable CREATE VIEW ON stmt with PIVOT

Disable CREATE VIEW ON stmt with PIVOT operator due to pg_dump cannot
dump pivot view with correct dependency and newly added fields in
FuncExpr casued readFunc failed to parse previous minor version string
parsetree during select from view after mVU.

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4673] Remove newly added pivot fields in fcinfo

Removed newly added pivot fields in fcinfo and set the fcinfo->context
to pivot data when the FuncExpr is a bbf_pivot function

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

xhfanhe commented on src/include/nodes/primnodes.h:
```diff
@@ -695,6 +695,10 @@ typedef struct FuncExpr
 	/* token location, or -1 if unknown */
 	int			location;
+	/* bbf_pivot function required rewritted parsetrees */
+	List 	   *pivot_parsetree pg_node_attr(query_jumble_ignore, read_write_ignore, read_as(NULL));
```
Please use a more general name like `extra_info` or `custom_list` so that other features can make use of this field as well. Same applies to other engine changes.

RIC06X replied:
Replaced `pivot_parsetree` and `pivot_extrainfo` with `Node * context` field.

xhfanhe commented on src/include/fmgr.h:
```diff
@@ -18,6 +18,7 @@
 #ifndef FMGR_H
 #define FMGR_H
 
+#include "nodes/pg_list.h"
```
Why do we need this?

RIC06X replied:
`fmgr.h` is a header file used by `execSRF.c`. And `List` hasn't been used in `execSRF.c` until we added our pivot logic. So we need to include this to `fmgr.h`

RIC06X committed:
[BABEL-4713] Clean up added fileds in FunCall/FuncExpr

Replaced pivot related fields with generic Node * field in FuncCall and
FuncExpr

Task: BABEL-4713, BABEL-4714
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4713] Move code to Babelfish hook function

Move executor pivot data passing code to Babelfish hook function.

Task: BABEL-4713, BABEL-4714
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>
